network/p2p/inspector/control_message_metrics.go

```diff
@@ -1,29 +1,67 @@
 package inspector
 
 import (
+	"fmt"
+
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
 
-	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/engine/common/worker"
 	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/mempool/queue"
+	"github.com/onflow/flow-go/module/metrics"
 	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/network/p2p/inspector/internal"
 )
 
 const (
+	// DefaultControlMsgMetricsInspectorNumberOfWorkers default number of workers for the inspector component.
+	DefaultControlMsgMetricsInspectorNumberOfWorkers = 1
+	// DefaultControlMsgMetricsInspectorQueueCacheSize is the default size of the message queue.
+	DefaultControlMsgMetricsInspectorQueueCacheSize = 100
 	// rpcInspectorComponentName the rpc inspector component name.
 	rpcInspectorComponentName = "gossipsub_rpc_metrics_observer_inspector"
 )
 
+// ObserveRPCMetricsRequest represents a request to capture metrics for the provided RPC
+type ObserveRPCMetricsRequest struct {
+	// Nonce adds random value so that when msg req is stored on hero store a unique ID can be created from the struct fields.
+	Nonce string
+	// From the sender of the RPC.
+	From peer.ID
+	// rpc the rpc message.
+	rpc *pubsub.RPC
+}
+
 // ControlMsgMetricsInspector a  GossipSub RPC inspector that will observe incoming RPC's and collect metrics related to control messages.
 type ControlMsgMetricsInspector struct {
```

Comment:
Since this module has been developed as a ReadyDoneAware component, I suggest enhancing its test coverage. We have already implemented a similar testing strategy for tracers, which serve a related purpose (see an example [here](https://github.com/onflow/flow-go/blob/c56427d9f324fa06f9d86ad7723f29851618da44/network/p2p/tracer/gossipSubMeshTracer_test.go#L31)). Thorough testing is crucial because any failure to function properly may manifest as issues in metrics collection. By ensuring comprehensive test coverage, we can identify potential problems early and maintain the overall reliability of the module.

Reply:
https://github.com/onflow/flow-go/pull/4132/commits/1b065d9abae4f9bebe68e096fd8d7ba8da674f19